File: app/src/main/kotlin/com/pleon/buyt/ui/PieChartView.java
package com.pleon.buyt.ui;

import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.PathMeasure;
import android.graphics.Point;
import android.graphics.RectF;
import android.util.AttributeSet;
import android.view.View;
import android.view.animation.LinearInterpolator;

import java.text.DecimalFormat;
import java.util.ArrayList;
import java.util.List;

import androidx.annotation.Nullable;

/**
 * Adopted from [https://github.com/luweibin3118/PieChartView]
 */
public class PieChartView extends View {

    private Paint mPaint = new Paint(Paint.ANTI_ALIAS_FLAG | Paint.DITHER_FLAG);
    private Path mPath = new Path(), drawLinePath = new Path();
    private PathMeasure mPathMeasure = new PathMeasure();
    private Canvas mCanvas;
    private int width, height;
    private RectF pieRectF = new RectF(), tempRectF = new RectF();
    private int radius;
    private List<Slice> sliceList = new ArrayList<>(), leftTypeList = new ArrayList<>(), rightTypeList = new ArrayList<>();
    private List<Point> itemPoints = new ArrayList<>();
    private int cell = 0;
    private float innerRadius = 0;
    private float offRadius = 0, offLine;
    private int textAlpha;
    private Point firstPoint;
    private int backGroundColor = 0xffffffff;
    private int itemTextSize = 30, textPadding = 8;
    private int defaultStartAngle = -90;
    private float pieCell;
    private ValueAnimator animator;
    private long animDuration = 1000;
    private Point startPoint = new Point();
    private Point centerPoint = new Point();
    private Point endPoint = new Point();
    private Point tempPoint = new Point();

    public PieChartView(Context context) {
        super(context);
    }

    public PieChartView(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
    }

//    @Override
//    protected void onAttachedToWindow() {
//        super.onAttachedToWindow();
//        startAnim();
//    }

//    @Override
//    protected void onDetachedFromWindow() {
//        super.onDetachedFromWindow();
//        if (animator != null) animator.cancel();
//    }

    public void startAnim() {
        animator = ValueAnimator.ofFloat(0, 360f * 2);
        animator.setDuration(animDuration);
        animator.setInterpolator(new LinearInterpolator());
        animator.addUpdateListener(animation -> {
            float value = (float) animation.getAnimatedValue();
            if (value < 360f) {
                offRadius = value;
                offLine = 0;
                textAlpha = 0;
            } else if (value >= 360f) {
                offRadius = 360f;
                offLine = (value - 360f) / 360f;
                if (offLine > 0.5f) {
                    textAlpha = (int) (255 * ((offLine - 0.5f) / 0.5f));
                } else {
                    textAlpha = 0;
                }
            } else if (value == 360f * 2) {
                offRadius = 360f;
                offLine = 1.0f;
                textAlpha = 255;
            }
            postInvalidate();
        });
        animator.start();
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldWid, int oldHei) {
        super.onSizeChanged(w, h, oldWid, oldHei);
        this.width = w;
        this.height = h;
        radius = (int) (Math.min(width, height) / 3.2);
        pieRectF.set(width / 2 - radius, height / 2 - radius, width / 2 + radius, height / 2 + radius);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        try {
            mCanvas = canvas;
            drawPie();
            if (offRadius == 360f) {
                drawTitle();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private void drawTitle() {
        resetPaint();
        float startRadius = defaultStartAngle;
        int count = rightTypeList.size();
        int h;
        if (count > 1) {
            h = (radius * 2) / (count - 1);
        } else {
            h = radius;
        }
        for (int i = 0; i < count; i++) {
            mPath.reset();
            Slice slice = rightTypeList.get(i);
            double angle = 2 * Math.PI * ((startRadius + slice.radius / 2) / 360d);
            int x = (int) (width / 2 + radius * Math.cos(angle));
            int y = (int) (height / 2 + radius * Math.sin(angle));
            startPoint.set(x, y);
            centerPoint.set((int) (width / 2 + radius * 1.2f), height / 2 - radius + h * (i));
            endPoint.set((int) (width * 0.98f), centerPoint.y);
            mPath.moveTo(startPoint.x, startPoint.y);
            mPath.lineTo(centerPoint.x, centerPoint.y);
            mPath.lineTo(endPoint.x, endPoint.y);
            resetPaint();
            mPaint.setStrokeWidth(2);
            mPaint.setColor(slice.color);
            mPaint.setStyle(Paint.Style.STROKE);
            mPathMeasure.setPath(mPath, false);
            drawLinePath.reset();
            mPathMeasure.getSegment(0, mPathMeasure.getLength() * offLine, drawLinePath, true);
            mCanvas.drawPath(drawLinePath, mPaint);
            startRadius += slice.radius;

            if (textAlpha > 0) {
                mPaint.setTextSize(itemTextSize);
                mPaint.setStyle(Paint.Style.FILL);
                mPaint.setTextAlign(Paint.Align.CENTER);
                mPaint.setAlpha(textAlpha);
                mCanvas.drawText(slice.type, centerPoint.x + (endPoint.x - centerPoint.x) / 2,
                        centerPoint.y - textPadding, mPaint);
                mPaint.setTextSize(itemTextSize * 0.85f);
                mCanvas.drawText(slice.getPercent(), centerPoint.x + (endPoint.x - centerPoint.x) / 2,
                        centerPoint.y + (itemTextSize + textPadding) * 0.8f, mPaint);
            }
        }

        count = leftTypeList.size();
        if (count > 1) {
            h = (radius * 2) / (count - 1);
        } else {
            h = radius;
        }

        for (int i = 0; i < count; i++) {
            mPath.reset();
            Slice slice = leftTypeList.get(i);
            double angle = 2 * Math.PI * ((startRadius + slice.radius / 2) / 360d);
            int x = (int) (width / 2 + radius * Math.cos(angle));
            int y = (int) (height / 2 + radius * Math.sin(angle));
            startPoint.set(x, y);
            centerPoint.set((int) (width / 2 - radius * 1.2f), height / 2 - radius + h * (count - 1 - i));
            endPoint.set((int) (width * 0.02f), centerPoint.y);
            mPath.moveTo(startPoint.x, startPoint.y);
            mPath.lineTo(centerPoint.x, centerPoint.y);
            mPath.lineTo(endPoint.x, endPoint.y);
            resetPaint();
            mPaint.setStrokeWidth(2);
            mPaint.setColor(slice.color);
            mPaint.setAntiAlias(true);
            mPaint.setDither(true);
            mPaint.setStyle(Paint.Style.STROKE);
            mPathMeasure.setPath(mPath, false);
            drawLinePath.reset();
            mPathMeasure.getSegment(0, mPathMeasure.getLength() * offLine, drawLinePath, true);
            mCanvas.drawPath(drawLinePath, mPaint);
            startRadius += slice.radius;

            if (textAlpha > 0) {
                mPaint.setTextSize(itemTextSize);
                mPaint.setStyle(Paint.Style.FILL);
                mPaint.setTextAlign(Paint.Align.CENTER);
                mPaint.setAlpha(textAlpha);
                mCanvas.drawText(slice.type, centerPoint.x + (endPoint.x - centerPoint.x) / 2,
                        centerPoint.y - textPadding, mPaint);
                mPaint.setTextSize(itemTextSize * 0.85f);
                mCanvas.drawText(slice.getPercent(), centerPoint.x + (endPoint.x - centerPoint.x) / 2,
                        centerPoint.y + (itemTextSize + textPadding) * 0.8f, mPaint);
            }
        }

        if (textAlpha == 1f) {
            sliceList.clear();
            leftTypeList.clear();
            rightTypeList.clear();
            itemPoints.clear();
        }
    }

    private void drawPie() {
        if (mCanvas == null) {
            return;
        }
        mCanvas.drawColor(backGroundColor);
        mPaint.setStyle(Paint.Style.FILL);
        int sum = 0;
        for (Slice slice : sliceList) {
            sum += slice.widget;
        }
        float a = 360f / sum;
        float startRadius = defaultStartAngle;
        float sumRadius = 0;
        leftTypeList.clear();
        rightTypeList.clear();
        itemPoints.clear();
        for (Slice slice : sliceList) {
            slice.radius = slice.widget * a;
            double al = 2 * Math.PI * ((startRadius + 90) / 360d);
            tempPoint.set((int) (width / 2 + radius * Math.sin(al)),
                    (int) (height / 2 - radius * Math.cos(al)));
            if (cell > 0) {
                if (startRadius == defaultStartAngle) {
                    firstPoint = tempPoint;
                }
            }

            double angle = 2 * Math.PI * ((startRadius + slice.radius / 2) / 360d);
            double sin = -Math.sin(angle);
            double cos = -Math.cos(angle);
            if (cos > 0) {
                leftTypeList.add(slice);
            } else {
                rightTypeList.add(slice);
            }
            sumRadius += Math.abs(slice.radius);
            mPaint.setStyle(Paint.Style.FILL);
            mPaint.setColor(slice.color);
            if (pieCell > 0) {
                if (sumRadius <= offRadius) {
                    tempRectF.set(pieRectF.left - (float) (pieCell * cos), pieRectF.top - (float) (pieCell * sin),
                            pieRectF.right - (float) (pieCell * cos), pieRectF.bottom - (float) (pieCell * sin));
                    mCanvas.drawArc(tempRectF, startRadius, slice.radius, true, mPaint);
                } else {
                    mCanvas.drawArc(tempRectF, startRadius, slice.radius - (Math.abs(offRadius - sumRadius)), true, mPaint);
                    break;
                }
            } else {
                if (sumRadius <= offRadius) {
                    mCanvas.drawArc(pieRectF, startRadius, slice.radius, true, mPaint);
                } else {
                    mCanvas.drawArc(pieRectF, startRadius, slice.radius - (Math.abs(offRadius - sumRadius)), true, mPaint);
                    break;
                }

            }
            startRadius += slice.radius;
            if (cell > 0 && pieCell == 0) {
                mPaint.setColor(backGroundColor);
                mPaint.setStrokeWidth(cell);
                mCanvas.drawLine(getWidth() / 2, getHeight() / 2, tempPoint.x, tempPoint.y, mPaint);
            }
        }
        if (cell > 0 && firstPoint != null && pieCell == 0) {
            mPaint.setColor(backGroundColor);
            mPaint.setStrokeWidth(cell);
            mCanvas.drawLine(getWidth() / 2, getHeight() / 2, firstPoint.x, firstPoint.y, mPaint);
        }
        mPaint.setStyle(Paint.Style.FILL);
        mPaint.setColor(backGroundColor);
        if (innerRadius > 0 && pieCell == 0) {
            mCanvas.drawCircle(width / 2, height / 2, radius * innerRadius, mPaint);
        }
    }

    public void resetPaint() {
        mPaint.reset();
        mPaint.setAntiAlias(true);
        mPaint.setDither(true);
        mPaint.setAlpha(256);
    }

    public void clearData() {
        sliceList.clear();
    }

    public void addSector(Slice slice) {
        sliceList.add(slice);
    }

    /**
     * Set the gap between slices.
     */
    public void setCell(int cell) {
        this.cell = cell;
    }

    public void setInnerRadius(float innerRadius) {
        if (innerRadius > 1f) innerRadius = 1f;
        else if (innerRadius < 0) innerRadius = 0;

        this.innerRadius = innerRadius;
    }

    public void setBackGroundColor(int backGroundColor) {
        this.backGroundColor = backGroundColor;
    }

    public void setItemTextSize(int itemTextSize) {
        this.itemTextSize = itemTextSize;
    }

    /**
     * Sets vertical padding between the text and its horizontal line
     */
    public void setTextPadding(int textPadding) {
        this.textPadding = textPadding;
    }

    public void setAnimDuration(long animDuration) {
        this.animDuration = animDuration;
    }

    public static class Slice {

        private static final DecimalFormat formatter = new DecimalFormat("0.0%");
        String type;
        int widget;
        int color;
        float radius;

        public Slice(String type, int widget, int color) {
            this.type = type;
            this.widget = widget;
            this.color = color;
        }

        String getPercent() {
            return formatter.format(radius / 360f);
        }
    }
}

File: app/src/main/kotlin/com/pleon/buyt/ui/PieChartView.kt
package com.pleon.buyt.ui

import android.animation.ValueAnimator
import android.content.Context
import android.graphics.*
import android.graphics.Paint.ANTI_ALIAS_FLAG
import android.graphics.Paint.DITHER_FLAG
import android.util.AttributeSet
import android.view.View
import android.view.animation.LinearInterpolator
import androidx.core.content.ContextCompat.getColor
import androidx.core.content.res.ResourcesCompat
import com.pleon.buyt.R
import com.pleon.buyt.util.formatPercent
import org.jetbrains.anko.dip
import org.jetbrains.anko.sp
import kotlin.math.*

/**
 * Adopted from [https://github.com/luweibin3118/PieChartView]
 */
class PieChartView : View {

    private val mPaint = Paint(ANTI_ALIAS_FLAG or DITHER_FLAG)
    private val mPath = Path()
    private val drawLinePath = Path()
    private val mPathMeasure = PathMeasure()
    private var mWidth: Int = 0
    private var mHeight: Int = 0
    private val pieRectF = RectF()
    private var radius: Int = 0
    private val slices = ArrayList<Slice>()
    private val leftTypeList = ArrayList<Slice>()
    private val rightTypeList = ArrayList<Slice>()
    private var gap = 4f
    private var innerRadius = 80f
    private var offRadius = 0f
    private var offLine = 0f
    private var textAlpha: Int = 0
    private var emptyColor = getColor(context, R.color.chartEmptyColor)
    private var itemTextSize = sp(13.5f)
    private var textPadding = 8
    private val defaultStartAngle = -90
    private val pieCell = 0f
    private var animator: ValueAnimator? = null
    private val startPoint = Point()
    private val centerPoint = Point()
    private val endPoint = Point()
    private val tempPoint = Point()

    constructor(cxt: Context) : super(cxt) {
        mPaint.typeface = ResourcesCompat.getFont(cxt, R.font.vazir_wol_v_26_0_2_scaled_uniformly_to_94_percent)
    }

    constructor(cxt: Context, attrs: AttributeSet?) : super(cxt, attrs) {
        mPaint.typeface = ResourcesCompat.getFont(cxt, R.font.vazir_wol_v_26_0_2_scaled_uniformly_to_94_percent)
    }

    fun startAnim(duration: Long) {
        animator = ValueAnimator.ofFloat(0f, 360f * 2)
        animator!!.duration = duration
        animator!!.interpolator = LinearInterpolator()
        animator!!.addUpdateListener { animation ->
            val value = animation.animatedValue as Float
            if (value < 360f) {
                offRadius = value
                offLine = 0f
                textAlpha = 0
            } else if (value >= 360f) {
                offRadius = 360f
                offLine = (value - 360f) / 360f
                textAlpha = if (offLine > 0.5f) (255 * ((offLine - 0.5f) / 0.5f)).toInt() else 0
            } else if (value == 360f * 2) {
                offRadius = 360f
                offLine = 1.0f
                textAlpha = 255
            }
            postInvalidate()
        }
        animator?.start()
    }

    override fun onSizeChanged(w: Int, h: Int, oldWid: Int, oldHei: Int) {
        super.onSizeChanged(w, h, oldWid, oldHei)
        mWidth = w
        mHeight = h
        radius = (min(mWidth, mHeight) / 3.2).toInt()
        pieRectF.set((mWidth / 2 - radius).toFloat(), (mHeight / 2 - radius).toFloat(), (mWidth / 2 + radius).toFloat(), (mHeight / 2 + radius).toFloat())
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        drawPie(canvas)
        if (offRadius == 360f) drawTitle(canvas)
    }

    private fun drawPie(canvas: Canvas) {
        // in order to be able to clear part of canvas (making hole in pie) these are required
        // for more info see [https://stackoverflow.com/q/19947835/8583692]
        val bitmap = Bitmap.createBitmap(canvas.width, canvas.height, Bitmap.Config.ARGB_8888)
        val temp = Canvas(bitmap)
        val transparentPaint = Paint(ANTI_ALIAS_FLAG)
        transparentPaint.xfermode = PorterDuffXfermode(PorterDuff.Mode.CLEAR)

        if (slices.isEmpty()) { // draw empty pie
            mPaint.style = Paint.Style.FILL
            mPaint.color = emptyColor
            temp.drawCircle(mWidth / 2f, mHeight / 2f, radius.toFloat(), mPaint)
        }

        val sum = slices.sumOf(Slice::weight).toFloat()

        var startRadius = defaultStartAngle.toFloat()
        var sumRadius = 0f
        leftTypeList.clear()
        rightTypeList.clear()
        for ((index, slice) in slices.withIndex()) {
            slice.radius = slice.weight.toFloat() / sum * 360f
            val al = 2.0 * PI * ((startRadius + 90) / 360.0)
            tempPoint.set((mWidth / 2 + radius * sin(al)).toInt(), (mHeight / 2 - radius * cos(al)).toInt())

            val angle = 2.0 * PI * ((startRadius + slice.radius / 2) / 360.0)
            val cos = -cos(angle)
            if (cos > 0) leftTypeList.add(slice) else rightTypeList.add(slice)

            sumRadius += abs(slice.radius)
            mPaint.style = Paint.Style.FILL
            mPaint.color = slice.color
            if (sumRadius <= offRadius) {
                temp.drawArc(pieRectF, startRadius, slice.radius, true, mPaint)
            } else {
                temp.drawArc(pieRectF, startRadius, slice.radius - abs(offRadius - sumRadius), true, mPaint)
                break
            }
            startRadius += slice.radius
            // draw transparent separating lines
            if (slices.size > 1 && gap > 0 && pieCell == 0f) {
                transparentPaint.strokeWidth = if (index == 0) gap * 2 else gap
                temp.drawLine(width / 2f, height / 2f, tempPoint.x.toFloat(), tempPoint.y.toFloat(), transparentPaint)
            }
        }

        // inner hole
        temp.drawCircle(mWidth / 2f, mHeight / 2f, innerRadius, transparentPaint)

        canvas.drawBitmap(bitmap, 0f, 0f, mPaint)
    }

    private fun drawTitle(canvas: Canvas) {
        resetPaint()

        var startRadius = defaultStartAngle.toFloat()
        var count = rightTypeList.size
        var h = if (count > 1) radius * 2 / (count - 1) else radius

        for (i in 0 until count) {
            mPath.reset()
            val slice = rightTypeList[i]
            val angle = 2.0 * PI * ((startRadius + slice.radius / 2) / 360.0)
            val x = (mWidth / 2 + radius * cos(angle)).toInt()
            val y = (mHeight / 2 + radius * sin(angle)).toInt()
            startPoint.set(x, y)
            centerPoint.set((mWidth / 2 + radius * 1.2f).toInt(), mHeight / 2 - radius + h * i)
            endPoint.set((mWidth * 0.98f).toInt(), centerPoint.y)
            mPath.moveTo(startPoint.x.toFloat(), startPoint.y.toFloat())
            mPath.lineTo(centerPoint.x.toFloat(), centerPoint.y.toFloat())
            mPath.lineTo(endPoint.x.toFloat(), endPoint.y.toFloat())
            resetPaint()
            mPaint.strokeWidth = dip(1.85f).toFloat()
            mPaint.color = slice.color
            mPaint.style = Paint.Style.STROKE
            mPathMeasure.setPath(mPath, false)
            drawLinePath.reset()
            mPathMeasure.getSegment(0f, mPathMeasure.length * offLine, drawLinePath, true)
            canvas.drawPath(drawLinePath, mPaint)
            startRadius += slice.radius

            if (textAlpha > 0) {
                mPaint.textSize = itemTextSize.toFloat()
                mPaint.style = Paint.Style.FILL
                mPaint.textAlign = Paint.Align.CENTER
                mPaint.alpha = textAlpha
                canvas.drawText(slice.type, (centerPoint.x + (endPoint.x - centerPoint.x) / 2).toFloat(),
                        centerPoint.y - (textPadding * 2.4f), mPaint)
                mPaint.textSize = itemTextSize * 0.85f
                canvas.drawText(slice.percent, (centerPoint.x + (endPoint.x - centerPoint.x) / 2).toFloat(),
                        centerPoint.y + (itemTextSize + textPadding) * 0.8f, mPaint)
            }
        }

        count = leftTypeList.size
        h = if (count > 1) radius * 2 / (count - 1) else radius

        for (i in 0 until count) {
            mPath.reset()
            val slice = leftTypeList[i]
            val angle = 2.0 * PI * ((startRadius + slice.radius / 2) / 360.0)
            val x = (mWidth / 2 + radius * cos(angle)).toInt()
            val y = (mHeight / 2 + radius * sin(angle)).toInt()
            startPoint.set(x, y)
            centerPoint.set((mWidth / 2 - radius * 1.2f).toInt(), mHeight / 2 - radius + h * (count - 1 - i))
            endPoint.set((mWidth * 0.02f).toInt(), centerPoint.y)
            mPath.moveTo(startPoint.x.toFloat(), startPoint.y.toFloat())
            mPath.lineTo(centerPoint.x.toFloat(), centerPoint.y.toFloat())
            mPath.lineTo(endPoint.x.toFloat(), endPoint.y.toFloat())
            resetPaint()
            mPaint.strokeWidth = dip(1.85f).toFloat()
            mPaint.color = slice.color
            mPaint.isAntiAlias = true
            mPaint.isDither = true
            mPaint.style = Paint.Style.STROKE
            mPathMeasure.setPath(mPath, false)
            drawLinePath.reset()
            mPathMeasure.getSegment(0f, mPathMeasure.length * offLine, drawLinePath, true)
            canvas.drawPath(drawLinePath, mPaint)
            startRadius += slice.radius

            if (textAlpha > 0) {
                mPaint.textSize = itemTextSize.toFloat()
                mPaint.style = Paint.Style.FILL
                mPaint.textAlign = Paint.Align.CENTER
                mPaint.alpha = textAlpha
                canvas.drawText(slice.type, (centerPoint.x + (endPoint.x - centerPoint.x) / 2).toFloat(),
                        centerPoint.y - (textPadding * 2.4f), mPaint)
                mPaint.textSize = itemTextSize * 0.85f
                canvas.drawText(slice.percent, (centerPoint.x + (endPoint.x - centerPoint.x) / 2).toFloat(),
                        centerPoint.y + (itemTextSize + textPadding) * 0.8f, mPaint)
            }
        }

        if (textAlpha.toFloat() == 1f) {
            slices.clear()
            leftTypeList.clear()
            rightTypeList.clear()
        }
    }

    private fun resetPaint() {
        mPaint.reset()
        mPaint.isAntiAlias = true
        mPaint.isDither = true
        mPaint.alpha = 256
    }

    fun clearData() = slices.clear()

    fun addSlice(slice: Slice) = slices.add(slice)

    fun setSliceGap(gap: Int) {
        this.gap = gap.toFloat()
    }

    fun setInnerRadius(innerRadius: Float) {
        this.innerRadius = innerRadius.coerceIn(0f, 1f)
    }

    fun setItemTextSize(itemTextSize: Float) {
        this.itemTextSize = sp(itemTextSize)
    }

    /**
     * Sets vertical padding between the text and its horizontal line
     */
    @Suppress("unused")
    fun setTextPadding(textPadding: Int) {
        this.textPadding = textPadding
    }

    class Slice(var type: String, var weight: Double, var color: Int) {
        var radius: Float = 0f
        val percent: String get() = formatPercent(radius / 360f)
    }
}
